File: AllInOne/src/Tree/BinaryTreeLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
import java.util.LinkedList;
import java.util.List;
/**
 * 
 * 102. Binary Tree Level Order Traversal
 * 
 * @author jingjiejiang
 * @history Mar 30, 2021
 * 
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        
        List<List<Integer>> resList = new LinkedList<>();

        buildLevel(root, resList, 1);

        return resList;
    }

    private void buildLevel(TreeNode root, List<List<Integer>> list, int depth) {

        if (root == null) return ;

        if (list.size() < depth) {
            list.add(new LinkedList<>());
        }

        list.get(depth - 1).add(root.val);

        buildLevel(root.left, list, depth + 1);
        buildLevel(root.right, list, depth + 1);
    }
}

File: AllInOne/src/Tree/BinaryTreeLevelOrderTraversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function levelOrder(root: TreeNode | null): number[][] {
  const res: number[][] = new Array();

  const buildLevelOrderArray = (
    root: TreeNode | null,
    res: number[][],
    depth: number
  ) => {
    if (!root) return;

    if (!res.length || res.length < depth) {
      res[depth - 1] = new Array();
    }

    res[depth - 1].push(root.val);

    buildLevelOrderArray(root.left, res, depth + 1);
    buildLevelOrderArray(root.right, res, depth + 1);
  };

  buildLevelOrderArray(root, res, 1);

  return res;
}
